core/src/main/scala/kafka/controller/ZookeeperClient.scala

```diff
@@ -67,54 +70,28 @@ class ZookeeperClient(connectString: String, sessionTimeoutMs: Int, connectionTi
-          countDownLatch.countDown()
-        }}, ctx)
-      case DeleteRequest(path, version, ctx) => zooKeeper.delete(path, version, new VoidCallback {
-        override def processResult(rc: Int, path: String, ctx: Any) = {
-          responseQueue.add(DeleteResponse(rc, path, ctx))
-          countDownLatch.countDown()
-        }}, ctx)
-      case ExistsRequest(path, ctx) => zooKeeper.exists(path, zNodeChangeHandlers.containsKey(path), new StatCallback {
-        override def processResult(rc: Int, path: String, ctx: Any, stat: Stat) = {
-          responseQueue.add(ExistsResponse(rc, path, ctx, stat))
-          countDownLatch.countDown()
-        }}, ctx)
-      case GetDataRequest(path, ctx) => zooKeeper.getData(path, zNodeChangeHandlers.containsKey(path), new DataCallback {
-        override def processResult(rc: Int, path: String, ctx: Any, data: Array[Byte], stat: Stat) = {
-          responseQueue.add(GetDataResponse(rc, path, ctx, data, stat))
-          countDownLatch.countDown()
-        }}, ctx)
-      case SetDataRequest(path, data, version, ctx) => zooKeeper.setData(path, data, version, new StatCallback {
-        override def processResult(rc: Int, path: String, ctx: Any, stat: Stat) = {
-          responseQueue.add(SetDataResponse(rc, path, ctx, stat))
-          countDownLatch.countDown()
-        }}, ctx)
-      case GetACLRequest(path, ctx) => zooKeeper.getACL(path, null, new ACLCallback {
-        override def processResult(rc: Int, path: String, ctx: Any, acl: java.util.List[ACL], stat: Stat): Unit = {
-          responseQueue.add(GetACLResponse(rc, path, ctx, Option(acl).map(_.asScala).orNull, stat))
-          countDownLatch.countDown()
-        }}, ctx)
-      case SetACLRequest(path, acl, version, ctx) => zooKeeper.setACL(path, acl.asJava, version, new StatCallback {
-        override def processResult(rc: Int, path: String, ctx: Any, stat: Stat) = {
-          responseQueue.add(SetACLResponse(rc, path, ctx, stat))
-          countDownLatch.countDown()
-        }}, ctx)
-      case GetChildrenRequest(path, ctx) => zooKeeper.getChildren(path, zNodeChildChangeHandlers.containsKey(path), new Children2Callback {
-        override def processResult(rc: Int, path: String, ctx: Any, children: java.util.List[String], stat: Stat) = {
-          responseQueue.add(GetChildrenResponse(rc, path, ctx, Option(children).map(_.asScala).orNull, stat))
-          countDownLatch.countDown()
-        }}, ctx)
+    val responseQueue = new ArrayBlockingQueue[Req#Response](requests.size)
+
+    requests.foreach { request =>
+      val processedRequest: Req = request match {
+        case r: WatchableAsyncRequest if r.shouldWatch.isEmpty =>
+          val shouldWatch = r match {
+            case _: GetChildrenRequest => zNodeChildChangeHandlers.contains(r.path)
+            case _: ExistsRequest | _: GetDataRequest => zNodeChangeHandlers.contains(r.path)
+          }
+          r.withWatch(Some(shouldWatch)).asInstanceOf[Req]
```

Comment:
It seems like we're creating copies of the user-provided requests just to set the watch flag. This is one area where I think the existing code had it right: just lookup the handler registrations as you're making the call to the underlying ZooKeeper client.

Reply:
We can avoid copies if important, but this hardly seems an issue compared to the number of copies being done elsewhere. For example, we do several collection copies on each iteration of `retryUntilConnected` in trunk.